File: frontend/src/Functions/Endpoints/Public/marketPricesQuery.test.js
import { afterEach, describe, expect, it, vi } from "vitest";

const fetchWithPublicHeaders = vi.fn();
vi.mock("./applyPublicHeaders.js", () => ({
  fetchWithPublicHeaders: (...args) => fetchWithPublicHeaders(...args),
}));

const { fetchMarketPricesQuery } = await import("./marketPricesQuery.js");

const ok = (body) => ({ ok: true, json: async () => body });
const want = (typeID, sourceID) => ({ typeID, sourceID });

afterEach(() => {
  fetchWithPublicHeaders.mockReset();
});

const bodyOf = (call) => JSON.parse(call[1].body);

describe("asking for prices", () => {
  it("names each market with the types wanted at it", async () => {
    fetchWithPublicHeaders.mockResolvedValue(
      ok({
        sources: { jita: { refreshedAt: 1, prices: { 34: { sell: 5 } } } },
      }),
    );

    const result = await fetchMarketPricesQuery({ wants: [want(34, "jita")] });

    expect(bodyOf(fetchWithPublicHeaders.mock.calls[0])).toEqual({
      sources: { jita: ["34"] },
      adjustedTypeIDs: [],
    });
    expect(result.sources.jita.prices["34"].sell).toBe(5);
  });

  // The whole point of the pair: a caller wanting two types at two markets is
  // not asking for four prices. One flat type list applied to every market
  // would fetch the two nobody reads.
  it("does not ask any market for a type wanted at another", async () => {
    fetchWithPublicHeaders.mockResolvedValue(ok({ sources: {} }));

    await fetchMarketPricesQuery({
      wants: [want(34, "jita"), want(35, "amarr")],
    });

    expect(bodyOf(fetchWithPublicHeaders.mock.calls[0]).sources).toEqual({
      jita: ["34"],
      amarr: ["35"],
    });
  });

  it("asks for a type at both markets when both are wanted", async () => {
    fetchWithPublicHeaders.mockResolvedValue(ok({ sources: {} }));

    await fetchMarketPricesQuery({
      wants: [want(34, "jita"), want(34, "amarr")],
    });

    expect(bodyOf(fetchWithPublicHeaders.mock.calls[0]).sources).toEqual({
      jita: ["34"],
      amarr: ["34"],
    });
  });

  // Adjusted prices belong to no market, so they travel in their own list and a
  // caller wanting only those still has a request to make.
  it("asks for adjusted prices without naming a market", async () => {
    fetchWithPublicHeaders.mockResolvedValue(ok({ sources: {} }));

    await fetchMarketPricesQuery({ wants: [], adjustedTypeIDs: [34] });

    expect(bodyOf(fetchWithPublicHeaders.mock.calls[0])).toEqual({
      sources: {},
      adjustedTypeIDs: ["34"],
    });
  });

  it("asks for nothing when nothing is wanted", async () => {
    await fetchMarketPricesQuery({ wants: [] });
    await fetchMarketPricesQuery({ wants: [want(34, "")] });

    expect(fetchWithPublicHeaders).not.toHaveBeenCalled();
  });

  it("asks about a repeated pair once", async () => {
    fetchWithPublicHeaders.mockResolvedValue(ok({ sources: {} }));

    await fetchMarketPricesQuery({
      wants: [want(34, "jita"), want(34, "jita"), want(35, "jita")],
    });

    expect(bodyOf(fetchWithPublicHeaders.mock.calls[0]).sources.jita).toEqual([
      "34",
      "35",
    ]);
  });
});

describe("more prices than one request may carry", () => {
  const manyWants = Array.from({ length: 750 }, (_, i) => want(i + 1, "jita"));

  const answerWith = () =>
    fetchWithPublicHeaders.mockImplementation(async (_url, options) => {
      const { sources, adjustedTypeIDs } = JSON.parse(options.body);
      return ok({
        sources: Object.fromEntries(
          Object.entries(sources).map(([id, typeIDs]) => [
            id,
            {
              refreshedAt: 1757000000000,
              prices: Object.fromEntries(
                typeIDs.map((typeID) => [typeID, { sell: 5 }]),
              ),
            },
          ]),
        ),
        adjusted: adjustedTypeIDs.length
          ? {
              refreshedAt: 1756900000000,
              prices: Object.fromEntries(
                adjustedTypeIDs.map((typeID) => [typeID, 4.9]),
              ),
            }
          : undefined,
      });
    });

  // The rows nest under each source, so folding two answers with one shallow
  // merge would replace the first chunk's source block whole and take every
  // price in it with it.
  it("keeps the rows from every chunk", async () => {
    answerWith();

    const result = await fetchMarketPricesQuery({ wants: manyWants });

    expect(fetchWithPublicHeaders).toHaveBeenCalledTimes(2);
    expect(Object.keys(result.sources.jita.prices)).toHaveLength(750);
    expect(result.sources.jita.prices["1"]).toBeDefined();
    expect(result.sources.jita.prices["750"]).toBeDefined();
    expect(result.sources.jita.refreshedAt).toBe(1757000000000);
  });

  it("keeps the adjusted prices from every chunk", async () => {
    answerWith();

    const result = await fetchMarketPricesQuery({
      wants: [],
      adjustedTypeIDs: manyWants.map(({ typeID }) => typeID),
    });

    expect(Object.keys(result.adjusted.prices)).toHaveLength(750);
  });

  // The cap counts reads, not types: the same type at two markets is two of
  // them, so a split that counted types would send a request the server refuses.
  it("counts one type at two markets as two of the request's places", async () => {
    answerWith();

    await fetchMarketPricesQuery({
      wants: [
        ...manyWants.slice(0, 300),
        ...manyWants.slice(0, 300).map(({ typeID }) => want(typeID, "amarr")),
      ],
    });

    for (const call of fetchWithPublicHeaders.mock.calls) {
      const { sources } = bodyOf(call);
      const asked = Object.values(sources).reduce(
        (total, ids) => total + ids.length,
        0,
      );
      expect(asked).toBeLessThanOrEqual(500);
    }
  });
});

describe("when the request cannot be answered", () => {
  // A market that answered and held no order, and a market that could not be
  // reached, are different facts. Settling the second as the first would have
  // the cache above tell a reader there is no price until the entry goes stale.
  it("throws on a refusal rather than answering empty", async () => {
    fetchWithPublicHeaders.mockResolvedValue({ ok: false, status: 503 });

    await expect(
      fetchMarketPricesQuery({ wants: [want(34, "jita")] }),
    ).rejects.toThrow(/503/);
  });

  it("lets a network failure through", async () => {
    fetchWithPublicHeaders.mockRejectedValue(new Error("offline"));

    await expect(
      fetchMarketPricesQuery({ wants: [want(34, "jita")] }),
    ).rejects.toThrow("offline");
  });
});

File: services/api/v1endpoints/marketPricesQuery_test.go
package v1endpoints

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strconv"
	"testing"
	"time"

	"eve-industry-planner/api/apideps"
	esitypes "eve-industry-planner/shared/core/esi/types"
	"eve-industry-planner/testing/redisfake"

	eipredis "eve-industry-planner/shared/redis"
)

const (
	jitaRegion  = 10000002
	amarrRegion = 10000043
)

func queryHandler(t *testing.T) (*Handlers, *eipredis.Redis) {
	t.Helper()
	redis := eipredis.NewRedis(redisfake.New(t).Client)
	return New(&apideps.Deps{Redis: redis}), redis
}

func query(t *testing.T, h *Handlers, body MarketPricesQueryBody) (*httptest.ResponseRecorder, MarketPricesQueryResponse) {
	t.Helper()
	encoded, err := json.Marshal(body)
	if err != nil {
		t.Fatalf("encode the request: %v", err)
	}
	rec := httptest.NewRecorder()
	h.MarketPricesQueryHandler(rec, httptest.NewRequest(
		http.MethodPost, "/api/v1/market-prices/query", bytes.NewReader(encoded)))

	var got MarketPricesQueryResponse
	if rec.Code == http.StatusOK {
		if err := json.Unmarshal(rec.Body.Bytes(), &got); err != nil {
			t.Fatalf("decode the response: %v (body %s)", err, rec.Body.String())
		}
	}
	return rec, got
}

func seedPrice(t *testing.T, redis *eipredis.Redis, typeID, regionID int32, buy, sell float64) {
	t.Helper()
	err := redis.MarketOrders().PutPrice(context.Background(), typeID, regionID, eipredis.MarketPriceEntry{
		Buy: buy, Sell: sell, BuyP95: buy, SellP05: sell, LastUpdated: 1757000000000,
	})
	if err != nil {
		t.Fatalf("seed a price: %v", err)
	}
}

// The caller names the markets it wants, so a surface pricing against one market
// carries one market's figures rather than four and a choice.
func TestMarketPricesQueryAnswersOnlyTheSourcesAsked(t *testing.T) {
	h, redis := queryHandler(t)
	seedPrice(t, redis, 34, jitaRegion, 5, 6)
	seedPrice(t, redis, 34, amarrRegion, 7, 8)

	_, got := query(t, h, MarketPricesQueryBody{Sources: map[string][]string{"jita": {"34"}}})

	if len(got.Sources) != 1 {
		t.Fatalf("got %d sources, want 1: %+v", len(got.Sources), got.Sources)
	}
	if price := got.Sources["jita"].Prices["34"]; price.Buy != 5 || price.Sell != 6 {
		t.Errorf("jita 34 = %+v, want buy 5 sell 6", price)
	}
}

// A market holding no order for a type says nothing about it, rather than
// reporting a price of zero — which is a figure, and a wrong one.
func TestMarketPricesQueryOmitsATypeAMarketHasNoOrderFor(t *testing.T) {
	h, redis := queryHandler(t)
	seedPrice(t, redis, 34, jitaRegion, 5, 6)

	_, got := query(t, h, MarketPricesQueryBody{
		Sources: map[string][]string{"jita": {"34", "35"}},
	})

	if _, held := got.Sources["jita"].Prices["35"]; held {
		t.Errorf("35 was answered for: %+v", got.Sources["jita"].Prices)
	}
	if _, held := got.Sources["jita"].Prices["34"]; !held {
		t.Error("34 was not answered for")
	}
}

// Each source carries the moment its region was walked, which is what lets a
// browser tell a row it already holds from one a walk has replaced.
func TestMarketPricesQueryCarriesEachSourcesClock(t *testing.T) {
	h, redis := queryHandler(t)
	walked := time.UnixMilli(1757000000000).UTC()
	if err := redis.MarketOrders().PutRefreshTime(context.Background(), jitaRegion, walked); err != nil {
		t.Fatalf("record the walk: %v", err)
	}
	seedPrice(t, redis, 34, jitaRegion, 5, 6)

	_, got := query(t, h, MarketPricesQueryBody{Sources: map[string][]string{"jita": {"34"}}})

	if got.Sources["jita"].RefreshedAt != walked.UnixMilli() {
		t.Errorf("refreshedAt = %d, want %d", got.Sources["jita"].RefreshedAt, walked.UnixMilli())
	}
}

// A source the server does not price is a client-side mistake — a reader-saved
// market is the browser's to fetch — so it is refused rather than answered
// empty, which would read as a market with no orders.
func TestMarketPricesQueryRefusesASourceTheServerDoesNotPrice(t *testing.T) {
	h, _ := queryHandler(t)

	rec, _ := query(t, h, MarketPricesQueryBody{
		Sources: map[string][]string{"jita": {"34"}, "some-citadel": {"34"}},
	})

	if rec.Code != http.StatusBadRequest {
		t.Fatalf("status = %d, want 400", rec.Code)
	}
}

func TestMarketPricesQueryRefusesARequestAskingForNothing(t *testing.T) {
	h, _ := queryHandler(t)

	rec, _ := query(t, h, MarketPricesQueryBody{})

	if rec.Code != http.StatusBadRequest {
		t.Fatalf("status = %d, want 400", rec.Code)
	}
}

// Each market carries the types wanted at it, so a caller pricing some materials
// at one market and some at another is not asking for every type at both. One
// shared list would fetch and return the rows nothing reads.
func TestMarketPricesQueryAsksEachSourceOnlyForItsOwnTypes(t *testing.T) {
	h, redis := queryHandler(t)
	seedPrice(t, redis, 34, jitaRegion, 5, 6)
	seedPrice(t, redis, 35, jitaRegion, 1, 2)
	seedPrice(t, redis, 35, amarrRegion, 7, 8)

	_, got := query(t, h, MarketPricesQueryBody{
		Sources: map[string][]string{"jita": {"34"}, "amarr": {"35"}},
	})

	if _, held := got.Sources["jita"].Prices["35"]; held {
		t.Errorf("jita answered for 35, which was wanted at amarr: %+v", got.Sources["jita"].Prices)
	}
	if _, held := got.Sources["jita"].Prices["34"]; !held {
		t.Error("jita did not answer for 34")
	}
	if _, held := got.Sources["amarr"].Prices["35"]; !held {
		t.Error("amarr did not answer for 35")
	}
}

// A market that resolved none of its caller's types asks for nothing, which is
// not a mistake — every type it would have covered was priced somewhere else.
func TestMarketPricesQuerySkipsASourceWantingNoTypes(t *testing.T) {
	h, redis := queryHandler(t)
	seedPrice(t, redis, 34, jitaRegion, 5, 6)

	rec, got := query(t, h, MarketPricesQueryBody{
		Sources: map[string][]string{"jita": {"34"}, "amarr": {}},
	})

	if rec.Code != http.StatusOK {
		t.Fatalf("status = %d, want 200", rec.Code)
	}
	if _, answered := got.Sources["amarr"]; answered {
		t.Errorf("amarr was answered for though nothing was wanted there: %+v", got.Sources)
	}
}

// The adjusted price is source-independent and refreshes daily, so it is its own
// block with its own clock and is only sent when asked for.
func TestMarketPricesQuerySendsAdjustedPricesOnlyWhenAsked(t *testing.T) {
	h, redis := queryHandler(t)
	seedPrice(t, redis, 34, jitaRegion, 5, 6)
	err := redis.Cache(eipredis.DatasetMarketPrices).PutEntry(context.Background(), int32(34),
		esitypes.AdjustedPrice{TypeID: 34, AdjustedPrice: 4.9, LastUpdated: 1756900000000})
	if err != nil {
		t.Fatalf("seed an adjusted price: %v", err)
	}

	_, without := query(t, h, MarketPricesQueryBody{Sources: map[string][]string{"jita": {"34"}}})
	if without.Adjusted != nil {
		t.Errorf("adjusted prices were sent unasked: %+v", without.Adjusted)
	}

	_, with := query(t, h, MarketPricesQueryBody{
		Sources: map[string][]string{"jita": {"34"}}, AdjustedTypeIDs: []string{"34"},
	})
	if with.Adjusted == nil {
		t.Fatal("adjusted prices were asked for and not sent")
	}
	if with.Adjusted.Prices["34"] != 4.9 {
		t.Errorf("adjusted 34 = %v, want 4.9", with.Adjusted.Prices["34"])
	}
	if with.Adjusted.RefreshedAt != 1756900000000 {
		t.Errorf("adjusted clock = %d, want 1756900000000", with.Adjusted.RefreshedAt)
	}
}

// Adjusted prices belong to no market, so a caller wanting only those names none
// and is still answered.
func TestMarketPricesQueryAnswersAdjustedPricesWithNoSource(t *testing.T) {
	h, redis := queryHandler(t)
	err := redis.Cache(eipredis.DatasetMarketPrices).PutEntry(context.Background(), int32(34),
		esitypes.AdjustedPrice{TypeID: 34, AdjustedPrice: 4.9, LastUpdated: 1756900000000})
	if err != nil {
		t.Fatalf("seed an adjusted price: %v", err)
	}

	rec, got := query(t, h, MarketPricesQueryBody{AdjustedTypeIDs: []string{"34"}})

	if rec.Code != http.StatusOK {
		t.Fatalf("status = %d, want 200", rec.Code)
	}
	if got.Adjusted == nil || got.Adjusted.Prices["34"] != 4.9 {
		t.Errorf("adjusted = %+v, want 34 at 4.9", got.Adjusted)
	}
}

// The cap counts the reads a request asks for, not the type ids it names: the
// same type at two markets is two reads, so counting ids would let a request
// through that asks for twice the work.
func TestMarketPricesQueryCountsOneTypeAtTwoMarketsTwice(t *testing.T) {
	h, _ := queryHandler(t)

	typeIDs := make([]string, 0, 300)
	for id := 34; id < 334; id++ {
		typeIDs = append(typeIDs, strconv.Itoa(id))
	}

	rec, _ := query(t, h, MarketPricesQueryBody{
		Sources: map[string][]string{"jita": typeIDs, "amarr": typeIDs},
	})

	if rec.Code != http.StatusBadRequest {
		t.Fatalf("status = %d, want 400 for 600 reads", rec.Code)
	}
}

func TestMarketPricesQueryRefusesAnythingButPost(t *testing.T) {
	h, _ := queryHandler(t)

	rec := httptest.NewRecorder()
	h.MarketPricesQueryHandler(rec, httptest.NewRequest(http.MethodGet, "/api/v1/market-prices/query", nil))

	if rec.Code != http.StatusMethodNotAllowed {
		t.Fatalf("status = %d, want 405", rec.Code)
	}
}
